TPA gravity factor (for wings)

limonspb opened the pull request:
Follow up for this PRs:
https://github.com/betaflight/betaflight/pull/13662
https://github.com/betaflight/betaflight/pull/13661

covered with `#ifdef USE_WING`

## Current problem
For wings the TPA argument should depend on wing speed relative to the air. (for quads it's just throttle)
High speed - less PIDs, low speed - more PIDs.
`lpf2(trottle)` is a good estimator for the speed, however it doesn't cover cases, for when wing glides down or climbs up at steep angles. In such cases gravity acts together or against throttle depending on the nose angle (angle between craft nose and horizon plane).
![image](https://github.com/user-attachments/assets/831d0f61-e2c2-43c3-928f-a5f768c28e5f)

When wing glides down it gradually gains speed under gravity, thus wing needs less PIDs.
When wing flies up, the speed is decreased thus more PIDs needed.
Currently in master:
`tpa argument = lpf2(throttle)`

This can give problems, for example, when wing flies straight down with zero throttle, it gains substantial speed. With the current approach PIDs are maxed out causing oscillations (because TPA argument is zero), if PID tune is tight.

## Solution
This PR updates it to:
`tpa argument = lpf2(throttle - K * sin(nose_angle))`

Two tune parameters are introduced:
`tpa_gravity_factor_min` and `tpa_gravity_factor_max`

Good start for tuning is something around:
`set tpa_gravity_factor_min = 5`
`set tpa_gravity_factor_max = 20`

Means that at zero throttle the gravity adds/subtracts up to 0.2 to/from TPA argument.
At 100% speed the gravity adds/subtracts up to 0.05 to/from TPA argument.
So if wing flies nose straight down for long enough with zero throttle, it will have PIDs as if it has 20% throttle.
If wing flies nose down 45 degrees with zero throttle long enough, it will have PIDs like at 20% * sin(45) throttle.

lpf2 on top of everything acts somewhat like inertia, having gradual accumulative effect.

## Tuning tips
Eyeball or log how fast wing can fly straight down with zero throttle.
Estimate how much throttle you need to get to that speed when fly horizontally.
This amount of throttle in % is exactly what `tpa_gravity_factor_max` is.

Log how fast wing can fly horizontally at full throttle, and how much it slows down when flying full throttle nose up. (or by how much is speeds up when full throttle nose down vs full throttle horizontally)
The difference in precents is `tpa_gravity_factor_min`.

Heavy wings that can glide down fast need more `tpa_gravity_factor_min/max`.
Low profile wings also need more `tpa_gravity_factor_min/max`.
Very fast wings need less `tpa_gravity_factor_min`, because at full throttle gravity doesn't affect the speed as much.
Very slow/underpowered wings need more `tpa_gravity_factor_min`, since some of them pretty much stop, even at full throttle, when nose is up.

## Example logs
Here is a log example.
[black_wing_tpa_gravity_on_and_off.BBL.txt](https://github.com/user-attachments/files/16319297/black_wing_tpa_gravity_on_and_off.BBL.txt)

The third log in the file - without gravity factor, and when wing flies nose down long enough with zero throttle, it starts oscillating:
![image](https://github.com/user-attachments/assets/2efd5830-ad91-4ee1-84f2-1e2ca9937921)

Second log in the file - with gravity factor, and oscillations are gone, because TPA argument is not zero in the nose down-zero throttle situation:
![image](https://github.com/user-attachments/assets/4b466875-aaa4-4357-89a4-8918d12fc391)

## General thoughts
GPS 3D speed for TPA argument might be good but only for when no wind. 10-20mph wind - wing can still fly, but this error in estimating TPA argument can be dramatic.

New approach is heuristic, but works very good, and has parameters that are simple to understand and tune with bbx.

## Debug values:
`set debug_mode = TPA`
- debug 0 - `tpaFactor * 1000` - final PIDs multiplier both for quads and wings.
- debug 1 - nose angle in degrees (for wings only).
- debug 2 - `noseAngleFactor * 1000` - addition to TPA argument for wings.
- debug 3 - `tpaArgument * 1000` - final TPA argument for wings.

## Next steps
- Introduce proper TPA curve for wings.
- Take into account control surfaces angles for TPA argument. Because, for example, violent rolls shortly reduce forward speed.
- Exclude throttle factor when disarmed (wing should be able to fly when disarmed also)
- Don't accumulate gravity factor before the first arm (probably?)

ledvinap reviewed (commented):
I'm a bit skeptical about tpa_gravity_factor_min / max, at least as implemented now.

`tpa argument = lpf2(throttle - K * sin(nose_angle))` is physically sound (with linear throttle/force in 0..1 range, K = 1/<thrust-to-weight ratio>). From this, thrust-to-weight ratio may be only necessary parameter.

With min/max factors, there is additional DoF useable for tuning. But IMO math implementation closer to physical model should be used.

(I'll try to look at the physics soon)

blckmn commented:
I feel you really need the relative airflow to give an accurate air speed. GPS will only give you ground speed.

I've not checked but are there any small pitot/static sensor systems available? And importantly affordable.

You could potentially build one with two barometers (if sensitive enough). One piped (metal straw?) and sealed from the forward position, the other measuring the static pressure.

ledvinap commented:
@blckmn: I don't think that accurate speed estimation is necessary, this only handles PID attenuation.

Pitot systems are quite common(https://www.team-blacksheep.com/products/product:4404) , for PX4 and ardupilot. Adding it to BF would be quite easy, but it is useful feature?

You probably need differential barometer. Absolute accuracy of affordable baros would be far too low for low-speed detection.

ledvinap commented:
@limonspb : 
From tuning description, assuming linear throttle and linear speed-to-throttle relationship, you get `tpa_gravity_factor_min`=`tpa_gravity_factor_max` = `1 / <thrust_to_weight_ratio>`. 

With t2w large enough, the math seems to do something reasonable (thrust range scaled to accommodate gravity (into 0..1 range), actual thrust has no effect; TPA=1 when flying straight down, TPA=0 when hanging on propeller).

With T2W == 1, the equation is problematic - tpa_gravity_factor_max=tpa_gravity_factor_min=1 and TPA is based only on pitch angle (TPA is limited to 0 with positive pith, TPA=1 when flying straight down).

limonspb commented:
> @limonspb : From tuning description, assuming linear throttle and linear speed-to-throttle relationship, you get `tpa_gravity_factor_min`=`tpa_gravity_factor_max` = `1 / <thrust_to_weight_ratio>`.
> 
> With t2w large enough, the math seems to do something reasonable (thrust range scaled to accommodate gravity (into 0..1 range), actual thrust has no effect; TPA=1 when flying straight down, TPA=0 when hanging on propeller).
> 
> With T2W == 1, the equation is problematic - tpa_gravity_factor_max=tpa_gravity_factor_min=1 and TPA is based only on pitch angle (TPA is limited to 0 with positive pith, TPA=1 when flying straight down).

First of all, just wanna say I really appreciate your reviews. They are always thorough and deep.

About the problem you mentioned... Let me think about it

limonspb commented:
> I feel you really need the relative airflow to give an accurate air speed. GPS will only give you ground speed.
> 
> I've not checked but are there any small pitot/static sensor systems available? And importantly affordable.
> 
> You could potentially build one with two barometers (if sensitive enough). One piped (metal straw?) and sealed from the forward position, the other measuring the static pressure.

Yeah the idea here is to make a bare minimum wing fly good in acro. So just accel, gyro, and throttle is used to estimate pid attenuation. And indeed GPS would work, but for non windy days. So I don't use it here as well.

limonspb commented:
> > I feel you really need the relative airflow to give an accurate air speed. GPS will only give you ground speed.
> > I've not checked but are there any small pitot/static sensor systems available? And importantly affordable.
> > You could potentially build one with two barometers (if sensitive enough). One piped (metal straw?) and sealed from the forward position, the other measuring the static pressure.
> 
> Yeah the idea here is to make a bare minimum wing fly good in acro. So just accel, gyro, and throttle is used to estimate pid attenuation. And indeed GPS would work, but for non windy days. So I don't use it here as well.

If someone in future can ass a support for the actual wind speed sensor, it be easy to just replace current estimation of TPA argument with the sensor values. Will just have to know the wing MAX speed.

limonspb commented:
> > > I feel you really need the relative airflow to give an accurate air speed. GPS will only give you ground speed.
> > > I've not checked but are there any small pitot/static sensor systems available? And importantly affordable.
> > > You could potentially build one with two barometers (if sensitive enough). One piped (metal straw?) and sealed from the forward position, the other measuring the static pressure.
> > 
> > 
> > Yeah the idea here is to make a bare minimum wing fly good in acro. So just accel, gyro, and throttle is used to estimate pid attenuation. And indeed GPS would work, but for non windy days. So I don't use it here as well.
> 
> If someone in future can ass a support for the actual wind speed sensor, it be easy to just replace current estimation of TPA argument with the sensor values. Will just have to know the wing MAX speed.

But i'd assume most people would fly without it anyways. Like most people fly without GPS :)

limonspb commented:
> @limonspb : From tuning description, assuming linear throttle and linear speed-to-throttle relationship, you get `tpa_gravity_factor_min`=`tpa_gravity_factor_max` = `1 / <thrust_to_weight_ratio>`.
> 
> With t2w large enough, the math seems to do something reasonable (thrust range scaled to accommodate gravity (into 0..1 range), actual thrust has no effect; TPA=1 when flying straight down, TPA=0 when hanging on propeller).
> 
> With T2W == 1, the equation is problematic - tpa_gravity_factor_max=tpa_gravity_factor_min=1 and TPA is based only on pitch angle (TPA is limited to 0 with positive pith, TPA=1 when flying straight down).

ok i see what you mean.
If `tpaGravityFactorMin == 1` then this line is a problem:
`throttle = scaleRangef(throttle, 0.0f, 1.0f, 0.0f, 1.0f - pidRuntime.tpaGravityFactorMin);` because it just rescales throttle to the range [0,0] making it zero all the time.

I think instead of this:
```
    const float noseAngleFactor = ....
    throttle = scaleRangef(throttle, 0.0f, 1.0f, 0.0f, 1.0f - pidRuntime.tpaGravityFactorMin);
    float tpaArgument = throttle - noseAngleFactor;
```
i need:
```
    const float noseAngleFactor = ....
    float tpaArgument = throttle - noseAngleFactor;
    tpaArgument = scaleRangef(tpaArgument, 0.0f, 1.0 + pidRuntime.tpaGravityFactorMax, 0.0f, 1.0f);
```

limonspb commented on src/main/flight/pid.c:
```diff
@@ -281,32 +283,59 @@ void pidResetIterm(void)
+
+    const float sinNoseAngle = getNoseAngleSin();
+    DEBUG_SET(DEBUG_TPA, 1, lrintf(asin_approx(sinNoseAngle) * 180.0 / M_PIf));
+    const float angleFactorAdjustment = pidRuntime.tpaGravityFactorMax + 
```
i think you are right, probably easier to just use `throttle` here to find a value between `tpaGravityFactorMax` and `tpaGravityFactorMin`

limonspb commented:
> I'm a bit skeptical about tpa_gravity_factor_min / max, at least as implemented now.
> 
> `tpa argument = lpf2(throttle - K * sin(nose_angle))` is physically sound (with linear throttle/force in 0..1 range, K = 1/). From this, thrust-to-weight ratio may be only necessary parameter.
> 
> With min/max factors, there is additional DoF useable for tuning. But IMO math implementation closer to physical model should be used.
> 
> (I'll try to look at the physics soon)

So initially i was thinking it should be just K parameter in CLI without min/max. But then i noticed that the faster wing goes the less gravity affects the final speed.

For my wing, for example:
Nose straight down, 0 throttle flies 20mph.
Full throttle horizontally it flies 100mph.
But i have to try really hard to get 110mph at full throttle nose down.

ledvinap commented:
>  `tpaArgument = scaleRangef(tpaArgument, 0.0f, 1.0 + pidRuntime.tpaGravityFactorMax, 0.0f, 1.0f);`

This should handle linear case well, even T/W < 1 ( tpaGravityFactorM?? > 1 ) will get bounded results.

> For my wing, for example:
Nose straight down, 0 throttle flies 20mph+.
Full throttle horizontally it flies 100mph.
But i have to try really hard to get 110mph at full throttle nose down.

I'm guessing a bit, but the cause is that you are in regime where total drag is roughly proportional to square of speed (parasitic drag dominates https://en.wikipedia.org/wiki/Parasitic_drag). So effect of gravity is large at slow speed and small at maximum speed. 
The situation may be different for 'classical' RC airplane with low thrust/weight ratio. But TPA sounds like something that is most useful for overpowered planes. 

Propeller thrust (at low airspeed relative to pitch speed)) is roughly proportional to square of RPM (https://www.electricrcaircraftguy.com/2013/09/propeller-static-dynamic-thrust-equation.html), result is linear part of your throttle/speed curve. At maximum throttle, it may be limited by airspeed or regulator/battery limits.


Thinking about it - user should specify linear and quadratic coefficient for speed/TPA estimation. But it is useless without some easy way to determine them (and guessing will do more harm than good)

limonspb commented:
> > `tpaArgument = scaleRangef(tpaArgument, 0.0f, 1.0 + pidRuntime.tpaGravityFactorMax, 0.0f, 1.0f);`
> 
> This should handle linear case well, even T/W < 1 ( tpaGravityFactorM?? > 1 ) will get bounded results.
> 
> > For my wing, for example:
> > Nose straight down, 0 throttle flies 20mph+.
> > Full throttle horizontally it flies 100mph.
> > But i have to try really hard to get 110mph at full throttle nose down.
> 
> I'm guessing a bit, but the cause is that you are in regime where total drag is roughly proportional to square of speed (parasitic drag dominates https://en.wikipedia.org/wiki/Parasitic_drag). So effect of gravity is large at slow speed and small at maximum speed. The situation may be different for 'classical' RC airplane with low thrust/weight ratio. But TPA sounds like something that is most useful for overpowered planes.
> 
> Propeller thrust (at low airspeed relative to pitch speed)) is roughly proportional to square of RPM (https://www.electricrcaircraftguy.com/2013/09/propeller-static-dynamic-thrust-equation.html), result is linear part of your throttle/speed curve. At maximum throttle, it may be limited by airspeed or regulator/battery limits.
> 
> Thinking about it - user should specify linear and quadratic coefficient for speed/TPA estimation. But it is useless without some easy way to determine them (and guessing will do more harm than good)

so im thinking now:
1. get rid of `tpa_gravity_factor_min`, and leave just K. Name it tpa_weight_trust_ratio.
2. In the next PR introduce TPA argument curve

limonspb commented on src/main/flight/imu.c:
```diff
@@ -165,6 +165,15 @@ STATIC_UNIT_TESTED void imuComputeRotationMatrix(void)
+#ifdef USE_WING
+// Angle in between nose axis of the craft and the horizontal plane in ground reference.
+// Poitive angle - nose up, negative angle - nose down.
+float getNoseAngleSin(void)
```
it's not exactly pitch, because normally pitch is in the craft coordinate system.
For example you can roll 90 degrees, and then pitch up and down. The "nose angle" will be zero. But pitch will be changing.

I'll make in consistent with `getCosTiltAngle`

ledvinap replied:
Not in betaflight. In body frame, rotations are applied in yaw/pitch/roll (around z-y'-x'' body axes) order. So with roll=90deg, elevator will change yaw, not pitch.

Also, in BF, positive pitch means nose down (https://github.com/betaflight/betaflight/blob/master/src/main/flight/imu.c#L317). It is better to stick to this convention. Coordinate systems are already complicated enough in BF.

sugaarK commented:
> I feel you really need the relative airflow to give an accurate air speed. GPS will only give you ground speed.
> 
> I've not checked but are there any small pitot/static sensor systems available? And importantly affordable.
> 
> You could potentially build one with two barometers (if sensitive enough). One piped (metal straw?) and sealed from the forward position, the other measuring the static pressure.

can get some sample pitot tubes sent out if its a worth while add

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -749,6 +749,13 @@ void imuUpdateAttitude(timeUs_t currentTimeUs)
 
+// Angle in between nose axis of the craft and the horizontal plane in ground reference.
+// Poitive angle - nose down, negative angle - nose up.
+float getSinTiltAngle(void)
```
`getCosTiltAngle` gets angle between Body Z and World up - both roll and pitch cause tilt. `getSinNoseUpAngle`? But that sounds terrible

limonspb replied:
oh yeah, they are different anles... How about "getSinDiveAngle"? Diving down - positive dive angle. Negative dive angle - not diving.

ctzsnooze replied:
@limon aren't you mostly concerned about pitch down?  If the wing was side-slipping, like roll down but (say) with no pitch down, in that case you'd not have much authority?  I am a little unsure about giving equal weight to the roll element in the tilt factor.

limonspb replied:
@ctzsnooze this was my bad to name `getSinTiltAngle`, because it actually returns the "dive" angle, basically how far the nose it down. Roll can be a factor, but not by itself. For example if you roll 90 degrees, then yaw will affect how far nose is down.

ctzsnooze replied:
Thanks for the clarification.
Maybe 'getSinNoseDownAngle` or just `getNoseDownAngle` might be OK :-)

haslinghuis replied:
Agree function name should not include technical implementation detail

limonspb replied:
i kinda like that "Dive angle" idea...
"getSinDiveAngle" sounds better than "getSinNoseDown"?

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -165,6 +165,15 @@ STATIC_UNIT_TESTED void imuComputeRotationMatrix(void)
+// Poitive angle - nose up, negative angle - nose down.
+float getNoseAngleSin(void)
+{
+    return rMat[2][0];
```
(the link should be to https://github.com/betaflight/betaflight/blob/master/src/main/flight/imu.c#L317)

ledvinap commented:
> * get rid of `tpa_gravity_factor_min`, and leave just K. Name it `tpa_weight_thrust_ratio`? Or `tpa_gravity_factor`?
> * In the next PR introduce TPA argument curve that will help for the wings that don't have linear speed(throttle). It will basically have one parameter "curavature" like that (picture), but most people, probably would not need to play with it. Unless they have a wing with a huge drag, or highly inefficient motors/props at high throttle.

(took some time to think about it)

This won't help. Thrust as function of RPM (and thus throttle) is quadratic, so thrust + gravity won't work (https://github.com/betaflight/betaflight/pull/13778#issuecomment-2246242510).

Maybe using `force-to-speed(throttle-to-force(throttle) + G * sin(alpha))` can work. `throttle-to-force` is quite well documented (but is speed-dependent). And `force-to-speed` can be approximated as curve you described (it is roughly `k*sqrt(force)` in high-power region and (IMO) k2*force in cruise conditions). (And it gets terribly complicated when you are at wrong Reynold's numbers: https://en.wikipedia.org/wiki/Drag_coefficient#/media/File:Drag_coefficient_on_a_sphere_vs._Reynolds_number_-_main_trends.svg).
At least it gives rough physical background to this problem.

Using your 2-point method (with new TPA range extension) is attractive idea - it is simple to explain, quite easy to measure (your Tuning Tips) and error may be acceptable (this method approximates parabola as line - it will underestimate gravity influence on mid throttle). (maybe the error can be easily reduced without complicating the algorithm too much).
Some things to consider:
- Naming is confusing - `tpa_gravity_factor_min/max` are now limits on gravity factors, but it is more related to min/max throttle. Meaning  of min/max is opposite in each context. Maybe `tpa_gravity_th0` `tpa_gravity_th90` or something similar?
- it may be better to tune high-throttle point in top of linear region (~90% throttle in your case). But only if simple method is possible
- use throttle directly to map tpa_factor (maybe the equations can be solved to get better thrust/equivalent throttle estimate)

limonspb commented:
> > * get rid of `tpa_gravity_factor_min`, and leave just K. Name it `tpa_weight_thrust_ratio`? Or `tpa_gravity_factor`?
> > * In the next PR introduce TPA argument curve that will help for the wings that don't have linear speed(throttle). It will basically have one parameter "curavature" like that (picture), but most people, probably would not need to play with it. Unless they have a wing with a huge drag, or highly inefficient motors/props at high throttle.
> 
> (took some time to think about it)
> 
> This won't help. Thrust as function of RPM (and thus throttle) is quadratic, so thrust + gravity won't work ([#13778 (comment)](https://github.com/betaflight/betaflight/pull/13778#issuecomment-2246242510)).
> 
> Maybe using `force-to-speed(throttle-to-force(throttle) + G * sin(alpha))` can work. `throttle-to-force` is quite well documented (but is speed-dependent). And `force-to-speed` can be approximated as curve you described (it is roughly `k*sqrt(force)` in high-power region and (IMO) k2*force in cruise conditions). (And it gets terribly complicated when you are at wrong Reynold's numbers: https://en.wikipedia.org/wiki/Drag_coefficient#/media/File:Drag_coefficient_on_a_sphere_vs._Reynolds_number_-_main_trends.svg). At least it gives rough physical background to this problem.
> 
> Using your 2-point method (with new TPA range extension) is attractive idea - it is simple to explain, quite easy to measure (your Tuning Tips) and error may be acceptable (this method approximates parabola as line - it will underestimate gravity influence on mid throttle). (maybe the error can be easily reduced without complicating the algorithm too much). Some things to consider:
> 
> * Naming is confusing - `tpa_gravity_factor_min/max` are now limits on gravity factors, but it is more related to min/max throttle. Meaning  of min/max is opposite in each context. Maybe `tpa_gravity_th0` `tpa_gravity_th90` or something similar?
> * it may be better to tune high-throttle point in top of linear region (~90% throttle in your case). But only if simple method is possible
> * use throttle directly to map tpa_factor (maybe the equations can be solved to get better thrust/equivalent throttle estimate)

Sounds good to make `tpa_gravity_th100` and `tpa_gravity_th0` instead of min/max and use just throttle to calculate the multiplier for `G*sin(alpha)`. Just thinking that `tpa_gravity_th90` might be a bit more confusing for tuning than `tpa_gravity_th100`.

And what you think about reusing thrust linearization we already have for the final formula:
`speed = lpf2(thrust_linearization(throttle) + K(throttle) * G * sin(alpha))`

limonspb commented:
@ledvinap i renamed min/max to thr100 and thr0 (i think it will be easier for users that way), renamed `getSinTilAngle` to `getSinDiveAngle`.
Was about to scale between thr100 and thr0 with just throttle, but figured it would be inaccurate. Because how much gravity affects in % really depends on "speed". Since we don't know the new speed, we can rely on the previous speed value. Throttle can change instantly from 0 to 100. If we use that, it will also instantly change `angleFactorAdjustment`.

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -749,6 +749,13 @@ void imuUpdateAttitude(timeUs_t currentTimeUs)
 #endif // USE_ACC
 
+// Angle in between the nose axis of the craft and the horizontal plane in ground reference.
+// Poitive angle - nose down, negative angle - nose up.
```
(positive)

Is sign of value correct? IIRC `rMat[?][0]` is projection of body X axis into earth frame, `rMat[2][0]` is Z component of it. With BF coordinate system, I'd expect it to be positive when pitching up (but BF coordinate system are sometimes a bit misleading).

Also, in BF, `pitch = asin(-rMat[2][0]);`. 
Because pitch is second angle, pitch and dive are equivalent (this won't be true for roll and projection of Y axis). `getSinPitchAngle` name will work in this case.

limonspb replied:
you are right, this is just pitch angle in IMU. I will double check the sign.
In this case, do you think it's better to just use `sin(attitude.values.pitch)` or it's worth to save a little performance, and still use `getSinPitchAngle()`?

ledvinap commented:
> Was about to scale between thr100 and thr0 with just throttle, but figured it would be more accurate to use throttle after lpf2 so that angleFactorAdjustment couldn't change instantly with throttle stick command.

IMO this is more complicated than necessary.

Physics approximation:

Assuming TPA is based on speed:
This kind of works because `thrust ~= k1 * throttle ^ 2` and `speed ~= sqrt(thrust)`. So `speed ~= k2 * throttle` (see comments above).

You can add gravity force to thrust (and use thrust-to-weight ratio as only parameter), but then you need either way to convert thrust to speed (sqrt is very rough approximation working only under some conditions) or base TPA on thrust and not speed.

Using `throttleLpf` to interpolate thr0..thr100:

'angleFactorAdjustment couldn't change instantly' IMO there is no problem with scaling with throttle directly. You are using dive angle to adjust force acting on plane, but acting on speed instead of force (using linear approximation of function similar to https://www.wolframalpha.com/input?i=plot+y%3Dsqrt%28x%5E2+%2B+a%29-x+where+a+%3D.3+for+x%3D0..1%2C). Force generated by propeller changes almost instantly with throttle. LPF filter handles thrust-to-speed conversion (emulates inertia). It is applied both to throttle and correction factor.

Using iterative form (feeding back throttle + dive correction as input) will apply different values that measured in tuning - assume t/w=2. zero throttle, flying down. thr0=50, thr100=22. 50% is applied in first, then applied value is decreased, using converged value somewhere in first half of [50..22] interval, instead of user-measured value. (It may actually be more useful value - but it is different from value specified by user and it;'s hard to reason about).

ledvinap commented:
> And what you think about reusing thrust linearization we already have for the final formula:
> speed = lpf2(thrust_linearization(throttle) + K(throttle) * sin(alpha))

Problem is that  `speed = sqrt(thrust_linearization(throttle))` at high speeds and really complicated otherwise. And thrust_linearization is IIRC good only for static thrust.

> speed = lpf2(throttle + K(throttle) * sin(alpha))
because on top of that there will be another heuristic anyways (different PR) - the actual TPA curve TPA(speed). (currently it's three connected lines, but should be something more smooth and asymptotic as speed grows)

That's exactly what I'm suggesting (with `K(throttle) = scaleRangef(throttle, 0, 1, thr0, thr100)`). It's big simplification, but useful one ..

limonspb commented:
> > And what you think about reusing thrust linearization we already have for the final formula:
> > speed = lpf2(thrust_linearization(throttle) + K(throttle) * sin(alpha))
> 
> Problem is that `speed = sqrt(thrust_linearization(throttle))` at high speeds and really complicated otherwise. And thrust_linearization is IIRC good only for static thrust.
> 
> > speed = lpf2(throttle + K(throttle) * sin(alpha))
> > because on top of that there will be another heuristic anyways (different PR) - the actual TPA curve TPA(speed). (currently it's three connected lines, but should be something more smooth and asymptotic as speed grows)
> 
> That's exactly what I'm suggesting (with `K(throttle) = scaleRangef(throttle, 0, 1, thr0, thr100)`). It's big simplification, but useful one ..

agreed, done.
noticed also I need:
```
    const float maxTpaArgument = MAX(1.0 + pidRuntime.tpaGravityThr100, pidRuntime.tpaGravityThr0);
    tpaArgument = scaleRangef(tpaArgument, 0.0f, maxTpaArgument, 0.0f, 1.0f);
```
since `thr0` in theory could be > 1.0.

"dive angle" is indeed jsut IMU pitch and is positive when nose down, so i've adjusted the function.

ledvinap commented on src/main/flight/pid.h:
```diff
@@ -261,6 +261,8 @@ typedef struct pidProfile_s {
     uint8_t ez_landing_limit;               // Maximum motor output when all sticks centred and throttle zero
     uint8_t ez_landing_speed;               // Speed below which motor output is limited
     uint16_t tpa_delay_ms;                  // TPA delay for fixed wings using pt2 filter (time constant)
+    uint8_t tpa_gravity_thr0;               // For wings: addition to tpa argument in % when zero throttle
```
Better put it at end of existing PG structure

ledvinap reviewed (approved):
Thanks for patience! It's nice, clean PR now

ledvinap commented:
Ad for IMU pitch:
I prefer using rotation matrix, thinking about it as projection of forward axis (or as projection of gravity vector into plane longitudinal axis). Rotation matrix is easy to visualize, has no singularities, is easy to use from more complicated calculations.

But when the value is equivalent to `sin(pitch)` (and this is possible for pitch only) , it should match in naming and sign to avoid confusion.

haslinghuis reviewed (approved):
`tpaArgument` is a bit generic. What if you need another argument. Was thinking about `tpaThrottle`

limonspb commented:
> Thanks for patience! It's nice, clean PR now

Thanks a bunch for your help! It's much better with all our changes now.
Would you mind to take a look at this discussion about TPA curve?
https://github.com/betaflight/betaflight/discussions/13786
I've been trying to find a right curve for couple of weeks now...

KarateBrot commented on src/main/flight/pid.c:
```diff
@@ -281,32 +283,48 @@ void pidResetIterm(void)
+    const float tpaArgument = throttle;
+#endif
+
+    bool isThrottlePastTpaLowBreakpoint = (tpaArgument < pidRuntime.tpaLowBreakpoint && pidRuntime.tpaLowBreakpoint > 0.01f) ? false : true;
```
I know this is not your code but let's take the opportunity to make this expression look nicer 😀
Instead of inverting the result, I inverted the expression directly (de Morgan's law)
```js
(x && y) ? false : true   ==   !(x && y)   ==   (!x || !y)
```
```suggestion
    bool isThrottlePastTpaLowBreakpoint = (tpaArgument >= pidRuntime.tpaLowBreakpoint) || (pidRuntime.tpaLowBreakpoint <= 0.01f);
```

limonspb replied:
lol i thought that was fixed long ago ))

KarateBrot commented on src/main/flight/pid.c:
```diff
@@ -281,32 +283,48 @@ void pidResetIterm(void)
+#ifdef USE_WING
+static float getWingTpaArgument(float throttle)
+{
+    const float sinPitchAngle = getSinPitchAngle();
+    const float pitchFactorAdjustment = scaleRangef(throttle, 0.0f, 1.0f, pidRuntime.tpaGravityThr0, pidRuntime.tpaGravityThr100);
+    const float pitchAngleFactor = sinPitchAngle * pitchFactorAdjustment;
```
You can get rid of the `sinPitchAngle` variable and get the value directly
```suggestion
    const float pitchFactorAdjustment = scaleRangef(throttle, 0.0f, 1.0f, pidRuntime.tpaGravityThr0, pidRuntime.tpaGravityThr100);
    const float pitchAngleFactor = getSinPitchAngle() * pitchFactorAdjustment;
```

limonspb replied:
i like one action per line, but ok :)

KarateBrot commented:
And finally a question about [these two CLI params](https://github.com/betaflight/betaflight/pull/13778/files#diff-1fca665faca91d927d814882e17ad63e404a95a51adc50b4b9aab0e74e160f72R1270-R1271):

Can we be sure no one ever needs a value bigger than 255? It might be enough for your wing but will it be enough for all wings in existence? Maybe we could change these two CLI parameters to 16 bit (and limit them to something like 5.000 or 10.000 for now). Just to be sure there is enough room for tuning all the wings in existence.

limonspb commented:
> And finally (because I forgot to ask in my review) a question about [these two CLI params](https://github.com/betaflight/betaflight/pull/13778/files#diff-1fca665faca91d927d814882e17ad63e404a95a51adc50b4b9aab0e74e160f72R1270-R1271):
> 
> Can we be sure no one ever needs a value bigger than 255? It might be enough for your wing but will it be enough for all wings in existence? Maybe we could change these two CLI parameters to 16 bit (and limit them to something like 5.000 or 10.000 for now). Just to be sure there is enough room for tuning all the wings in existence.

This is a good point, but can't think of any plane that would dive down with zero throttle 2.56 times faster than it can fly horizontally. But maybe?

If Betaflight would be installed on big commercial airline, we would need like thr0 = 90 or 100?

limonspb commented:
> And finally (because I forgot to ask in my review) a question about [these two CLI params](https://github.com/betaflight/betaflight/pull/13778/files#diff-1fca665faca91d927d814882e17ad63e404a95a51adc50b4b9aab0e74e160f72R1270-R1271):
> 
> Can we be sure no one ever needs a value bigger than 255? It might be enough for your wing but will it be enough for all wings in existence? Maybe we could change these two CLI parameters to 16 bit (and limit them to something like 5.000 or 10.000 for now). Just to be sure there is enough room for tuning all the wings in existence.

done

ledvinap commented:
> This is a good point, but can't think of any plane that would dive down with zero throttle 2.56 times faster than it can fly horizontally. But maybe?

try converting foam toy glider, using small dc motors ;-) lots of fun, 4km/h cruise speed and really strange physics...

KarateBrot commented:
> can't think of any plane that would dive down with zero throttle 2.56 times faster than it can fly horizontally

I thought about it for a few minutes... To (dive) 2.55x as fast as you can travel, you need at least 16x (= 2.55^3) the power to overcome air friction (because `F ~ v²` and `P = F*v`).
```
P_dive / P_travel > 2.55^3
W * v_dive / (T * v_travel) > 2.55^3
W * 2.55 * v_travel /  (T * v_travel) > 2.55^3
W  /  T > 2.55^2
T/W < 1 / 2.55^2
```
This means a wing with a thrust-to-weight ratio smaller than 0.15 (= 1/2.55^2) might need a tunable value over 255, assuming quadratic air friction.

On [Wikipedia](https://en.wikipedia.org/wiki/Thrust-to-weight_ratio#Aircraft_2) the worst case scenario for a T-W ratio of a plane is listed with a T-W ratio of around 0.2 (B-2 bomber). So in the worst case scenario of this flying brick, the tuning parameter won't need to be bigger than around 220-230.

On the other hand, if the B-2 was perfectly streamlined and there was only laminar flow (linear air drag), the T/W condition would become
```
T/W < 1 / 2.55
```
and therefore the thrust-to-weight ratio only needs to be smaller than 0.4 to need a tunable value over 255. For the B-2 this yields the most extreme tuning value of around 500.
 
I guess the answer is somewhere in the middle, closer to 255. So you are probably right and 255 is plenty to work with, but we probably won't see values over 500, ever.

limonspb commented:
> > can't think of any plane that would dive down with zero throttle 2.56 times faster than it can fly horizontally
> 
> I thought about it for a few minutes... To (dive) 2.55x as fast as you can travel, you need at least 16x (= 2.55^3) the power to overcome air friction (because `F ~ v²` and `P = F*v`).
> 
> ```
> P_dive / P_travel > 2.55^3
> W * v_dive / (T * v_travel) > 2.55^3
> W * 2.55 * v_travel /  (T * v_travel) > 2.55^3
> W  /  T > 2.55^2
> 
> T/W < 1 / 2.55^2
> ```
> 
> This means a wing with a thrust-to-weight ratio smaller than 0.15 (= 1/2.55^2) might need a tunable value over 255, assuming quadratic air friction.
> 
> On [Wikipedia](https://en.wikipedia.org/wiki/Thrust-to-weight_ratio#Aircraft_2) the worst case scenario for a T-W ratio of a plane is listed with a T-W ratio of around 0.2 (B-2 bomber). So in the worst case scenario of this flying brick, the tuning parameter won't need to be bigger than around 220-230.
> 
> On the other hand, if the B-2 was perfectly streamlined and there was only laminar flow (linear air drag), the T/W condition would become
> 
> ```
> T/W < 1 / 2.55
> ```
> 
> and therefore the thrust-to-weight ratio only needs to be smaller than 0.4 to need a tunable value over 255. For the B-2 this yields the (unrealistic) tuning value of around 500. This is over-idealized and pushing the bounds of realism intentionally.
> 
> **I guess the answer is somewhere in the middle. So you are probably right and 255 is plenty to work with for most use cases. I wonder if we can actually get bigger values than 255, if we try to push it with a heavy and streamlined wing that has a low T-W ratio.**

I just made them UINT16 with max value of 5000, which is 5000% :) Opening the opportunity for super underpowered planes, like @ledvinap likes :)
There are gliders that don't even have throttle, so there's got to be "planes" with almost zero thrust.

Guys, could you plz have a look at this [discussion](https://github.com/betaflight/betaflight/discussions/13786) :)

KarateBrot commented on src/main/flight/pid.h:
```diff
@@ -72,6 +72,7 @@
 
 #ifdef USE_WING
 #define TPA_LOW_RATE_MIN INT8_MIN
+#define TPA_GRAVITY_MAX 5000
```
Now that I understand what this value does, I was probably a bit optimistic with such a high value 😀 1000 should be plenty.

ctzsnooze reviewed (approved):
Thanks @Limon for all your hard work here, and to @ledvinap and @KarateBrot for your contributions.  Great work!

haslinghuis commented:
@limonspb please update usage of debug values. Is this correct?

```
            'debug[0]': 'TPA Factor',
            'debug[1]': 'TPA Pitch Angle Factor (Wing)',
            'debug[2]': 'TPA Argument (Wing)',
```

limonspb commented:
> @limonspb please update usage of debug values. Is this correct?
> 
> ```
>             'debug[0]': 'TPA Factor',
>             'debug[1]': 'TPA Pitch Angle Factor (Wing)',
>             'debug[2]': 'TPA Argument (Wing)',
> ```

Yes sir, that's correct.
